### ISSUE TITLE
Metal executor doesn't wire LiteRtTopKMetalSampler_SetInputTensorsAndInferenceFunc — on-GPU sampling unreachable on macOS at v0.12.0

### ISSUE THREAD

**Issue by cipher-rc5** [google-ai-edge/LiteRT-LM#2502]
## Summary

On macOS arm64 with the Metal executor selected, every token sample falls back to CPU. The runtime correctly picks the Metal adapter for compute, but the TopK Metal sampler reports UNAVAILABLE at `Create()` time and the factory then falls through the WebGPU sampler (also unavailable) down to the CPU sampler. Net effect: per-token softmax+sample appears to round-trip Metal → CPU → Metal, and on a downstream wrapper we observe only ~1.3–1.6× GPU/CPU decode speedup on Apple M3 Ultra — far below what Apple Silicon should produce on a 4B-class model.

The prebuilt Metal sampler dylib does export the function we believe the executor is supposed to call to wire the sampler to the logit MTLBuffer (`LiteRtTopKMetalSampler_SetInputTensorsAndInferenceFunc`), but as far as we can tell from the public v0.12.0 surface, nothing in the executor calls it, and there is no C-ABI setter for downstream code to call it from outside the runtime. We'd like to confirm the diagnosis and offer to help with a fix.

## Environment

- **macOS:** 26.5.1
- **Host arch:** arm64
- **Xcode:** Xcode 26.5, Build version 17F42
- **Hardware:** Apple M3 Ultra, 96 GB RAM
- **LiteRT-LM:** v0.12.0 (commit `ffed38adbc33509480b5340e5173638bc20a68ff`)
- **Build configuration:** macOS arm64 slice produced by [`scripts/build_macos_xcframework.sh`](https://github.com/cipher-rc5/LiteRTLM-macOS/blob/main/scripts/build_macos_xcframework.sh). The script stages the Metal accelerator and sampler sidecars (`libLiteRtMetalAccelerator.dylib`, `libLiteRtTopKMetalSampler.dylib`, `libLiteRtWebGpuAccelerator.dylib`, `libLiteRtTopKWebGpuSampler.dylib`, `libLiteRt.dylib`) from `prebuilt/macos_arm64/` alongside the engine. The wrapper repo is public at https://github.com/cipher-rc5/LiteRTLM-macOS.
- **Model:** `litert-community/gemma-4-E4B-it-litert-lm` (3.4 GB), snapshot `f7ad3343bd6ebc9607f4dc3bc4f2398bd5749bc5`.

## Symptom

The runtime selects the Metal adapter correctly:

```text
adapter: Apple M3 Ultra, arch=metal-3, vendor=apple, backend=Metal, adapterType=Integrated GPU
Selected adapter: Apple M3 Ultra ...
```

But sampler initialization reports the following sequence in the verbose ABSL log:

```text
INFO: [gpu_environment.cc:364] Failed to create OpenCL context.
W0000 ... sampler_factory.cc:448] WebGPU sampler not available, falling back to statically linked C API: UNAVAILABLE: ERROR: [runtime/components/sampler_factory.cc:309]
W0000 ... sampler_factory.cc:728] GPU sampler unavailable. Falling back to CPU sampling. To use GPU sampling, please make sure libLiteRtTopKWebGpuSampler.so or libLiteRtTopKOpenClSampler.so is available at LD_LIBRARY_PATH on device.
```

The dylibs referenced in that final warning *are* present and `dlopen`-able — the Metal one in particular is staged next to the engine and its symbols load cleanly. The factory still falls through to CPU sampling.

## Diagnosis

The Metal sampler dylib exports the function we'd expect the executor to use to register logit tensors. From `nm -gU` against the staged dylib in our xcframework:

```text
$ nm -gU Frameworks/LiteRTLM.xcframework/macos-arm64/libLiteRtTopKMetalSampler.dylib
000000000000905c T _LiteRtTopKMetalSampler_CanHandleInput
0000000000007e04 T _LiteRtTopKMetalSampler_Create
0000000000008428 T _LiteRtTopKMetalSampler_Destroy
000000000000907c T _LiteRtTopKMetalSampler_HandlesInput
000000000000843c T _LiteRtTopKMetalSampler_SampleToIdAndScoreBuffer
000000000000909c T _LiteRtTopKMetalSampler_SetInputTensorsAndInferenceFunc
0000000000008cdc T _LiteRtTopKMetalSampler_UpdateConfig
```

Our reading is:

1. The Metal sampler's `Create()` requires that input logit tensors and the inference closure be registered via `_SetInputTensorsAndInferenceFunc` before it will accept itself as "available". [unverified — based on the symbol surface and the factory log pattern; we have not seen sampler source.]
2. The v0.12.0 Metal executor never calls `_SetInputTensorsAndInferenceFunc`. [unverified — based on the consistent UNAVAILABLE return and no public-API path that would let downstream code wire it.]
3. The factory then tries the WebGPU sampler against Metal-backed tensors, which also returns UNAVAILABLE (the `sampler_factory.cc:448` line above).
4. The factory falls through to CPU sampling (`sampler_factory.cc:728`), so each token's softmax/topK reads logits back from a Metal buffer, samples on CPU, and the next decode step writes the chosen ID back into a Metal buffer.

If steps 1–2 are correct, the per-token Metal→CPU→Metal sampling round-trip is unavoidable at v0.12.0 through any public API.

## Performance impact

On Apple M3 Ultra (96 GB) running Gemma 4 E4B through the wrapper at https://github.com/cipher-rc5/LiteRTLM-macOS, decode-only throughput is:

| backend | prompt | mean prefill_ms | mean decode_ms | mean tok/s (decode) |
|---------|--------|-----------------|----------------|---------------------|
| cpu     | short  | 536             | 104            | 28.8                |
| cpu     | medium | 537             | 4435           | 28.9                |
| cpu     | long   | 545             | 4816           | 26.7                |
| gpu     | short  | 104             | 3368           | 38.1                |
| gpu     | medium | 102             | 3044           | 42.2                |
| gpu     | long   |  98             | 2912           | 44.0                |

(5 iters + warmup per cell; top-p 0.95, temp 1.0; `max-output-tokens=128`, `max-num-tokens=1024`.)

For comparison, the [`litert-community/gemma-4-E2B-it-litert-lm` HF model card](https://huggingface.co/litert-community/gemma-4-E2B-it-litert-lm) reports **160.2 tok/s GPU decode on M4 Max** for the smaller E2B variant. Even allowing for the model-size difference, M3 Ultra at 38–44 tok/s on E4B is well below what Apple Silicon's memory bandwidth and Metal compute should produce. The per-token CPU sampling round-trip is our leading hypothesis for the gap — prefill is healthy on Metal (~5× faster than CPU), so the deficit is concentrated in the decode loop.

## Reproducer

1. Clone https://github.com/cipher-rc5/LiteRTLM-macOS (pinned at LiteRT-LM v0.12.0).
2. Build the macOS arm64 xcframework: `scripts/build_macos_xcframework.sh`. The script stages the five Metal/WebGPU sidecar dylibs from `prebuilt/macos_arm64/` into the xcframework slice.
3. Build the bench target: `swift build -c release --product BenchRunner`.
4. Download a Gemma 4 `.litertlm` (e.g. `litert-community/gemma-4-E4B-it-litert-lm`).
5. Run with verbose ABSL logs:
   ```sh
   ABSL_MIN_LOG_LEVEL=0 GLOG_v=2 \
   LITERTLM_MODEL_PATH=/path/to/gemma-4-E4B-it.litertlm \
     ./.build/release/BenchRunner --backend gpu
   ```
6. Observe the `sampler_factory.cc:448` and `sampler_factory.cc:728` log lines quoted above.

## What we believe the fix is

In the Metal executor's decode loop, before the first sample call, register the logit `MTLBuffer` (and the inference-step closure, if that's what the second argument expects) with the Metal sampler via `LiteRtTopKMetalSampler_SetInputTensorsAndInferenceFunc`. The Metal sampler dylib already exports this entry point, and the factory's UNAVAILABLE path is consistent with `Create()` refusing to construct until the inputs are registered. [unverified — we don't have visibility into the sampler's internal state machine; this is our best read from the public symbol surface and the factory log pattern.]

## Workaround

None that we've found through the public C ABI at v0.12.0. The samplers' setter is not exposed by the engine's public headers, so downstream wrappers can't reach it. `LITERT_LM_FST_CONSTRAINTS_DISABLED` and the other documented environment knobs don't affect this code path.

## Asks

1. **Confirmation of the diagnosis** from someone with visibility into the Metal executor source. In particular: is `_SetInputTensorsAndInferenceFunc` supposed to be wired up by the executor at decode-loop init time? Is there a different intended path we're missing?
2. **Status / ETA on a fix**, or a pointer to any in-progress work we could track. Happy to test internal builds against the M3 Ultra hardware here.
3. **Whether a downstream patch would be welcome**. If you'd accept a PR, we'd be willing to maintain a fork against v0.12.0 in the wrapper repo above and submit the fix upstream once we're confident it's the right shape. We'd appreciate a steer on the Metal-buffer ownership model first — that's the piece we'd need to understand from source to write the patch correctly.

Thanks for the work on LiteRT-LM — the Metal prefill path is already excellent, and closing the sampler gap would substantially raise the Apple Silicon decode ceiling.

## How to submit

The first line of this file is the issue title; strip it before submitting (or pass `--title` explicitly as below).

```sh
gh issue create \
  --repo google-ai-edge/LiteRT-LM \
  --title "Metal executor doesn't wire LiteRtTopKMetalSampler_SetInputTensorsAndInferenceFunc — on-GPU sampling unreachable on macOS at v0.12.0" \
  --body-file _dev/upstream/2026-06-03-metal-sampler-issue.md
```

If you'd prefer the body without the duplicate title line:

```sh
tail -n +3 _dev/upstream/2026-06-03-metal-sampler-issue.md > /tmp/issue-body.md
gh issue create \
  --repo google-ai-edge/LiteRT-LM \
  --title "Metal executor doesn't wire LiteRtTopKMetalSampler_SetInputTensorsAndInferenceFunc — on-GPU sampling unreachable on macOS at v0.12.0" \
  --body-file /tmp/issue-body.md
```

## Additional diagnostic evidence (Track C.1)

Captured 2026-06-03 from the staged xcframework at https://github.com/cipher-rc5/LiteRTLM-macOS — see `_dev/diagnostics/2026-06-03-metal-sampler-diag.md` for the full transcript.

**Binary-string evidence in `libLiteRTLMEngine.dylib`.** Running `strings` against the engine dylib produced from upstream sources at `ffed38a` shows the literal error strings that only appear on the suspected failure path inside `sampler_factory.cc`:

- `Metal sampler not available, falling back to other sampler options.`
- `SetInputTensors is not implemented.`
- `Could not load symbol`

These strings cannot be hit on the happy path. Their presence in the engine binary is consistent with the factory entering — and falling through — the Metal-sampler branch on every decode.

**Symbol asymmetry between engine and sampler.** `nm -gU` against the staged Metal sampler dylib (`libLiteRtTopKMetalSampler.dylib`) shows `_LiteRtTopKMetalSampler_SetInputTensorsAndInferenceFunc` exported (the function we believe the executor is supposed to call). The same `nm -gU` against `libLiteRTLMEngine.dylib` shows only the `_Static` weak-stub variant (`_…_SetInputTensorsAndInferenceFunc_Static`) plus a dlsym lookup string. The strong symbol is reachable via dlsym but, per the binary-string evidence above, the executor reports `SetInputTensors is not implemented` rather than wiring it.

**Methodological caveat.** The wrapper's `BenchRunner` calls `LogLevel.setMinimum(.error)` before engine init, which overrides `ABSL_MIN_LOG_LEVEL=0` at runtime. The captured `/tmp/sampler-diag.log` therefore shows only the adapter-selection lines (`adapter: Apple M3 Ultra, arch=metal-3, ..., backend=Metal`) and four WebGPU delegate init lines — no sampler warnings. We are deriving the failure path from the binary string set and the persistent Phase 5 decode-tok/s baseline (42.5 tok/s GPU, 4.2 tok/s CPU as the regression floor — both stable across 30 measured iterations in `Benchmarks/results/baseline-macos-2026-06-03.csv`).

If a developer with executor-source visibility can confirm whether `sampler_factory.cc:309 → :448 → :728` is the actual path being taken under `Backend.gpu` on macOS arm64, that would close the diagnosis loop. We can rebuild with a sampler-factory-internal `LOG(INFO)` patch if helpful for confirmation — happy to send a reproducer.

**Comment by cipher-rc5** [comment]
## Correction to the original diagnosis

After a deeper read of v0.12.0 sources, the framing in the original report ("the executor never calls `LiteRtTopKMetalSampler_SetInputTensorsAndInferenceFunc`") is incorrect. The executor *does* call it — the wiring is just gated on a predicate that doesn't fire on the Metal backend. Posting the corrected analysis so the issue tracks reality rather than our first read.

### What's actually happening

`runtime/executor/llm_litert_compiled_model_executor.cc` has the only consumer call site for the setter:

- `SetSamplerInputHandling()` at `:1298–:1313` is the wrapper that calls `sampler_->SetInputTensorsAndInferenceFunc(...)`.
- It's invoked from `InitializeSampler` at `:1279–:1280` — but only when `sampler_handles_input_` is true.
- That flag is set at `:1259–:1262`:

```cpp
sampler_handles_input_ =
    (!executor_settings_.GetAdvancedSettings().has_value() ||
     executor_settings_.GetAdvancedSettings()->sampler_handles_input) &&
    sampler_->CanHandleInput() && !signatures_.input_tokens.empty();
```

So the setter runs iff `sampler_->CanHandleInput()` returns `true` *at sampler init time*. For the Metal path that's the symbol `LiteRtTopKMetalSampler_CanHandleInput` in the prebuilt sampler dylib. The CPU-fallback path we observe (`sampler_factory.cc:728`) is consistent with `sampler_handles_input_` evaluating false because `CanHandleInput()` returned 0 at init — at which point the executor never even tries the wiring.

### Why this matters for the fix

Two possibilities, both worth confirming from someone with executor + sampler source visibility:

1. **The Metal sampler's `CanHandleInput` is supposed to return `true` unconditionally on macOS**, mirroring the WebGPU sampler (whose `Create` body at `sampler_factory.cc:412–:470` mirrors Metal's at `:504–:548` almost line-for-line, and whose `metal_exported_symbols.lds` symbol set matches WebGPU). If so, the bug is in the prebuilt `libLiteRtTopKMetalSampler.dylib` — the C-ABI surface is right but `CanHandleInput` is returning 0 when it shouldn't.

2. **The contract is "wire first, then `CanHandleInput` flips"** — i.e. the sampler is supposed to be told its input tensors *before* it can report it handles them. In that case the executor's gating order is reversed and the right fix is in LiteRT-LM: drop the `CanHandleInput()` predicate from the `sampler_handles_input_` evaluation (or attempt the wiring opportunistically and promote based on `HandlesInput()` afterward).

The Metal sampler dylib is closed-prebuilt for us, so we can't tell which contract `CanHandleInput` is honoring without source visibility.

### Asks (refined)

The asks in the original post still stand, but more precisely:

1. **Which contract is intended?** Is `CanHandleInput()` supposed to be `true` from Metal-sampler `Create()` time, or only after `SetInputTensorsAndInferenceFunc` has been called?
2. **If the contract is "true from Create()":** is the prebuilt `libLiteRtTopKMetalSampler.dylib` shipping a `CanHandleInput` body that returns 0? (We can confirm empirically by disassembling the symbol — it's ~32 bytes at offset `0x905c` in the dylib we have — but we'd rather hear it from someone with the source.)
3. **If the contract is "true after wiring":** would you accept a small patch to `llm_litert_compiled_model_executor.cc:1259–:1283` that attempts the wiring even when `CanHandleInput()` returns false, swallows `kUnimplemented`, and promotes `sampler_handles_input_` based on a subsequent `HandlesInput()` check? Happy to write it against v0.12.0 and PR.

Apologies for the noise in the original framing — wanted to correct it before anyone spent time chasing a phantom "executor never calls the setter" bug. The underlying observation (CPU fallback on Metal, ~44 tok/s decode ceiling on M3 Ultra) is unchanged.

**Comment by cipher-rc5** [comment]
## Closing as duplicate of #2073

Live ABSL log capture (with our wrapper's verbose-log suppression removed) shows the Metal-sampler hypothesis in this issue is **not** what's happening on the macOS GPU path. The real failure on Apple Silicon is the one already reported in #2073: the dispatcher takes the WebGPU branch, `dlsym(LiteRtTopKWebGpuSampler_UpdateConfig)` fails because the prebuilt `libLiteRtTopKWebGpuSampler.dylib` ships only 3 of the 7 C-ABI symbols `sampler_factory.cc` expects, and the chain falls through to CPU sampling. The Metal sampler dylib is never consulted.

### What actually fires on `Backend::GPU` on M3 Ultra / macOS 26.5 with v0.12.0

```
sampler_factory.cc:448  W  WebGPU sampler not available, falling back to statically linked C API: UNAVAILABLE: ERROR: [runtime/components/sampler_factory.cc:309]
                           └ Could not load symbol LiteRtTopKWebGpuSampler_UpdateConfig: dlsym(...): symbol not found.
sampler_factory.cc:685  I  WebGPU sampler not available, falling back to other sampler options.
sampler_factory.cc:728  W  GPU sampler unavailable. Falling back to CPU sampling.
```

No `Metal sampler not available` line. No `Dynamically loaded LiteRtTopKMetalSampler` line. The Metal-sampler branch isn't reached.

### Symbol set confirms #2073's audit for v0.12.0

```
$ nm -gU Frameworks/LiteRTLM.xcframework/macos-arm64/libLiteRtTopKWebGpuSampler.dylib | grep LiteRtTopK
T _LiteRtTopKWebGpuSampler_Create
T _LiteRtTopKWebGpuSampler_Destroy
T _LiteRtTopKWebGpuSampler_SampleToIdAndScoreBuffer
# 3/7 — same gap as v0.10.2

$ nm -gU Frameworks/LiteRTLM.xcframework/macos-arm64/libLiteRtTopKMetalSampler.dylib | grep LiteRtTopK
T _LiteRtTopKMetalSampler_CanHandleInput
T _LiteRtTopKMetalSampler_Create
T _LiteRtTopKMetalSampler_Destroy
T _LiteRtTopKMetalSampler_HandlesInput
T _LiteRtTopKMetalSampler_SampleToIdAndScoreBuffer
T _LiteRtTopKMetalSampler_SetInputTensorsAndInferenceFunc
T _LiteRtTopKMetalSampler_UpdateConfig
# 7/7 — the Metal sampler dylib is fine; the Metal branch is just never tried.
```

So the Metal sampler dylib is healthy in our build (we re-staged from the v0.12.0 dynamic-link Bazel artifacts), the WebGPU one isn't, and the dispatcher on macOS picks the WebGPU branch first. Fixing the WebGPU dylib's exports (per #2073) closes this on macOS without any LiteRT-LM-side change.

Cross-posting the same evidence on #2073 with our M3 Ultra repro for additional weight. Closing this issue as a duplicate. Apologies for the noise — our initial diagnosis chased binary-string evidence that only appears on a code path we never actually reach in this configuration.

Thanks to @DenisovAV for the original report.

**Comment by cipher-rc5** [comment]
Closing as duplicate of #2073 — see prior comment for evidence.
